Plugins/Public/mobiledocking_plugin/Main.cpp

```diff
@@ -201,252 +82,618 @@ void LoadSettings()
-	mobiledockClients[iClientID].iDockingModulesAvailable = (mobiledockClients[iClientID].iDockingModulesInstalled - mobiledockClients[iClientID].mapDockedShips.size());
+	// The client is free to dock, erase from the pending list and handle
+	mapPendingDockingRequests.erase(dockingID);
 
-	// If this is a ship which is currently docked, clear the market
-	if (mobiledockClients[iClientID].mobileDocked)
+	if (mobiledockClients[carrierID].iDockingModulesAvailable == 0)
 	{
-		SendResetMarketOverride(iClientID);
+		PrintUserCmdText(dockingID, L"Carrier has no free docking capacity");
+		return;
 	}
 
-}
-
-// Temporary storage for client data to be handled in LaunchPosHook.
-CLIENT_DATA undockingShip;
+	string scProxyBase = HkGetPlayerSystemS(carrierID) + "_proxy_base";
+	uint iBaseID;
+	Universe::IBase* baseInfo;
+	if (targetMobileDockingBase)
+	{
+		baseInfo = Universe::get_base(targetMobileDockingBase);
+		if (baseInfo == nullptr)
+		{
+			PrintUserCmdText(carrierID, L"Mobile Docking base not found, contact staff");
+			return;
+		}
+		iBaseID = targetMobileDockingBase;
+	}
+	else
+	{
+		string scProxyBase = HkGetPlayerSystemS(carrierID) + "_proxy_base";
+		if (pub::GetBaseID(iBaseID, scProxyBase.c_str()) == -4)
+		{
+			PrintUserCmdText(carrierID, L"No proxy base in system detected, contact staff. Required base: %ls", stows(scProxyBase).c_str());
+			return;
+		}
+		baseInfo = Universe::get_base(iBaseID);
+	}
 
-void __stdcall PlayerLaunch(unsigned int iShip, unsigned int client)
-{
+	// Save the carrier info
+	const wchar_t* carrierName = (const wchar_t*)Players.GetActiveCharacterName(carrierID);
+	const wchar_t* dockedName = (const wchar_t*)Players.GetActiveCharacterName(dockingID);
 
-	returncode = DEFAULT_RETURNCODE;
+	nameToDockedInfoMap[dockedName].carrierName = carrierName;
+	nameToDockedInfoMap[dockedName].lastDockedSolar = Players[dockingID].iLastBaseID;
+	nameToCarrierInfoMap[carrierName].dockedShipList.push_back(dockedName);
+	nameToCarrierInfoMap[carrierName].lastCarrierLogin = time(0);
+	idToCarrierInfoMap[carrierID] = &nameToCarrierInfoMap[carrierName];
+	idToDockedInfoMap[dockingID] = &nameToDockedInfoMap[dockedName];
 
-	uint carrier_client = HkGetClientIdFromCharname(mobiledockClients[client].wscDockedWithCharname);
+	mobiledockClients[carrierID].iDockingModulesAvailable--;
 
-	wstring clientName = (const wchar_t*)Players.GetActiveCharacterName(client);
+	uint playerSystem = Players[dockingID].iSystemID;
 
-	if (mobiledockClients[client].mobileDocked)
+	// Land the ship on the designated base
+	pub::Player::ForceLand(dockingID, iBaseID);
+	if (playerSystem != baseInfo->iSystemID)
 	{
+		Server.BaseEnter(baseInfo->iBaseID, dockingID);
+		Server.BaseExit(baseInfo->iBaseID, dockingID);
+		wstring wscCharFileName;
+		HkGetCharFileName(dockedName, wscCharFileName);
+		wscCharFileName += L".fl";
+		CHARACTER_ID cID;
+		strcpy(cID.szCharFilename, wstos(wscCharFileName.substr(0, 14)).c_str());
+		Server.CharacterSelect(cID, dockingID);

```

Comment:
This same block of code for base switching is used countless times. Make a helper method.